### PR TITLE
Add all master metrics

### PR DESCRIPTION
**Pull request by bobrik**
It feels dirty. Maybe we should generate this list from a few maps instead.

``` go
type mesosMetricMapping struct {
  metric string
  help   string
}

masterGauges := map[string]mesosMetricMapping{
 "mesos_master_cpus_revocable_total":  {"master/cpus_revocable_total", "Total number of revocable CPUs."},
 // ...
}

masterCounters := map[string]mesosMetricMapping{
  // ...
}
```

Metrics for messages look like they should be one metric with labels. Should I rewrite that?

Let me know what you think.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -82,6 +82,672 @@ var (
 	masterMetrics = []snapshotMetric{
 		snapshotMetric{
 			desc: prometheus.NewDesc(
+				"mesos_master_cpus_percent",
+				"Percent of CPUs used as a fraction of 1.",
```
Comment (brian-brazil) [inline main.go]: Is it a fraction or a percentage?

We prefer ratios, as it avoids confusion. The raw values are even better.

File: main.go [hunk 2]
```diff
@@ -82,6 +82,672 @@ var (
...
+				"mesos_master_cpus_revocable_percent",
+				"Percent of revocable CPUs used as a fraction of 1.",
+				masterMetricsLabels, nil,
+			),
+			snapshotKey: "master/cpus_revocable_percent",
+			valueType:   prometheus.GaugeValue,
+		},
+		snapshotMetric{
+			desc: prometheus.NewDesc(
+				"mesos_master_cpus_revocable_total",
```
Comment (brian-brazil) [inline main.go]: Gauges shouldn't end in _total, that's for counters

File: main.go [hunk 3]
```diff
@@ -82,6 +82,672 @@ var (
...
+				"Event queue dispatches.",
+				masterMetricsLabels, nil,
+			),
+			snapshotKey: "master/event_queue_dispatches",
+			valueType:   prometheus.CounterValue,
+		},
+		snapshotMetric{
+			desc: prometheus.NewDesc(
+				"mesos_master_event_queue_http_requests",
+				"Event queue HTTP requests.",
```
Comment (brian-brazil) [inline main.go]: This sounds like a gauge, can you make the description clearer?

File: main.go [hunk 4]
```diff
@@ -82,6 +82,672 @@ var (
...
+				masterMetricsLabels, nil,
+			),
+			snapshotKey: "master/event_queue_http_requests",
+			valueType:   prometheus.CounterValue,
+		},
+		snapshotMetric{
+			desc: prometheus.NewDesc(
+				"mesos_master_event_queue_messages",
+				"Event queue messages.",
```
Comment (brian-brazil) [inline main.go]: Processed or queued?

File: main.go [hunk 5]
```diff
@@ -82,6 +82,672 @@ var (
...
+				"mesos_master_frameworks_inactive",
+				"Number of inactive frameworks.",
+				masterMetricsLabels, nil,
+			),
+			snapshotKey: "master/frameworks_inactive",
+			valueType:   prometheus.GaugeValue,
+		},
+		snapshotMetric{
+			desc: prometheus.NewDesc(
+				"mesos_master_invalid_executor_to_framework_messages",
```
Comment (brian-brazil) [inline main.go]: Counters should end in _total

File: main.go [hunk 6]
```diff
@@ -82,6 +82,672 @@ var (
...
+				"Number of slave shutdowns completed.",
+				masterMetricsLabels, nil,
+			),
+			snapshotKey: "master/slave_shutdowns_completed",
+			valueType:   prometheus.CounterValue,
+		},
+		snapshotMetric{
+			desc: prometheus.NewDesc(
+				"mesos_master_slave_shutdowns_scheduled",
+				"Number of slave shutdowns scheduled.",
```
Comment (brian-brazil) [inline main.go]: Is this a counter or a gauge?

File: main.go [hunk 7]
```diff
@@ -152,6 +854,51 @@ var (
...
 			valueType:   prometheus.GaugeValue,
 		},
+		snapshotMetric{
+			desc: prometheus.NewDesc(
+				"mesos_master_uptime_secs",
+				"Mesos master uptime in seconds.",
+				masterMetricsLabels, nil,
+			),
+			snapshotKey: "master/uptime_secs",
+			valueType:   prometheus.CounterValue,
```
Comment (brian-brazil) [inline main.go]: This is actually a Gauge (you care about the absolute value, not how fast it's increasing)

File: main.go [hunk 8]
```diff
@@ -152,6 +854,51 @@ var (
 			snapshotKey: "master/tasks_starting",
 			valueType:   prometheus.GaugeValue,
 		},
+		snapshotMetric{
+			desc: prometheus.NewDesc(
+				"mesos_master_uptime_secs",
```
Comment (brian-brazil) [inline main.go]: "seconds" for consistency

**Comment by brian-brazil** [comment]
Someone with a bit more mesos knowledge should also take a look, I'm a bit rusty. @discordianfish ?

**Comment by brian-brazil** [comment]
> Metrics for messages look like they should be one metric with labels. Should I rewrite that?

That'd probably make sense.

File: main.go (hunk 1 above)
Comment (bobrik) [inline main.go]: I've tried to keep original names from Mesos stats. Do we care about that?

I can fix description in any case.

**Comment by bobrik** [comment]
Just found http://mesos.apache.org/documentation/latest/monitoring/, it makes sense to copy descriptions from there.

File: main.go (hunk 1 above)
Comment (brian-brazil) [inline main.go]: If you're going to exhaustively list every single metric, you may aswell fix them while you're at it.

Comment (bobrik) [inline main.go]: What do you mean by fixing? I'm still not sure if we try to keep metric names as close to mesos names as possible.

Comment (brian-brazil) [inline main.go]: In this sort of situation you'd try to keep them similar, and if there's big differences you can always put the original name in the help.

**Comment by bobrik** [comment]
Updated, PTAL. It seems that messages are too differen to put them into the same key.

File: main.go [hunk 9]
```diff
@@ -79,11 +79,669 @@ var (
...
+				"mesos_master_task_failed_source_slave_reason_container_launch_failed_total",
+				"Number of tasks failed because container launch failed on slave.",
+				masterMetricsLabels, nil,
+			),
+			snapshotKey: "master/task_failed/source_slave/reason_container_launch_failed",
+			valueType:   prometheus.CounterValue,
+		},
+		snapshotMetric{
+			desc: prometheus.NewDesc(
+				"mesos_master_task_failed_source_slave_reason_executor_terminated_total",
```
Comment (brian-brazil) [inline main.go]: The lost tasks reasons sounds like it could be one metric

File: main.go [hunk 10]
```diff
@@ -79,11 +79,669 @@ var (
...
 
+	// http://mesos.apache.org/documentation/latest/monitoring/
 	masterMetrics = []snapshotMetric{
 		snapshotMetric{
 			desc: prometheus.NewDesc(
+				"mesos_master_cpus_fraction",
+				"Fraction of allocated CPUs.",
+				masterMetricsLabels, nil,
+			),
+			snapshotKey: "master/cpus_percent",
```
Comment (brian-brazil) [inline main.go:125]: Does this need to be scaled, or am I right in reading the docs that Mesos has confused percent and ratio?

File: main.go [hunk 11]
```diff
@@ -79,11 +79,669 @@ var (
...
+				"mesos_master_disk_fraction",
+				"Fraction of allocated disk space.",
+				masterMetricsLabels, nil,
+			),
+			snapshotKey: "master/disk_percent",
+			valueType:   prometheus.GaugeValue,
+		},
+		snapshotMetric{
+			desc: prometheus.NewDesc(
+				"mesos_master_disk_revocable_fraction",
```
Comment (brian-brazil) [inline main.go:184]: We have the absolute values below, so we can drop this metric.

File: main.go [hunk 12]
```diff
@@ -79,11 +79,669 @@ var (
...
+				"mesos_master_disk_revocable",
+				"Revocable disk space in MB.",
+				masterMetricsLabels, nil,
+			),
+			snapshotKey: "master/disk_revocable_total",
+			valueType:   prometheus.GaugeValue,
+		},
+		snapshotMetric{
+			desc: prometheus.NewDesc(
+				"mesos_master_disk_revocable_used",
```
Comment (brian-brazil) [inline main.go]: The unit should be in the metric name, bytes is preferred

File: main.go (hunk 10 above)
Comment (bobrik) [inline main.go:125]: Mesos docs are wrong:

```
  "master/cpus_percent": 0.480022831050229,
  "master/cpus_total": 3504,
  "master/cpus_used": 1682,
```

File: main.go (hunk 11 above)
Comment (bobrik) [inline main.go:184]: Same deal as with memory and CPU. Do we want to drop all of them?

**Comment by bobrik** [comment]
Combined failure reasons and added number of tasks per framework. Can you take another look?

**Comment by discordianfish** [comment]
@bobrik: See #18 - You should probably base any changes upon https://github.com/mesosphere/mesos_exporter, we want to deprecate this repo. Does that work for you?

@brian-brazil / @juliusv I need permissions to the junkyard org to move this repo there.

**Comment by bobrik** [comment]
D'oh

**Comment by discordianfish** [comment]
I know, I know it sucks to waste time :( I'm sorry, I should have move this repo already. Will do as soon as I have the permissions. I hope the mesosphere exporter includes everything you need. Either way, it should be a much nicer base to work with.

**Comment by brian-brazil** [comment]
I guess this PR is obsolete then.

**Pull request closed without merging**